guides/unit-testing.md

```diff
@@ -0,0 +1,973 @@
+2. **Exercise:** Executing the code under test
+3. **Verify:** Confirming that the code behaves in an expected manner
+4. **Teardown:** Returning the environment to a clean slate
+
+Some phases may be represented in different forms, so it is important to be aware of their presence. For instance, the "setup" phase may occur within `beforeEach` hook, and the "teardown" phase may occur within an `afterEach` hook (or automatically by Jest).
+
+💡 It can help readability, especially with large tests, to demarcate phases with line breaks:
+
+1️⃣
+
+``` typescript
+describe('KeyringController', () => {
+  describe('cacheEncryptionKey', () => {
+    it('unlocks the keyrings with valid information', async () => {
+      const keyringController = await initializeKeyringController({
+        password: 'password',
+      });
+      keyringController.cacheEncryptionKey = true;
+      const returnValue = await keyringController.encryptor.decryptWithKey();
+      const decryptWithKeyStub = sinon.stub(
+        keyringController.encryptor,
+        'decryptWithKey',
+      );
+      decryptWithKeyStub.resolves(Promise.resolve(returnValue));
+
+      keyringController.store.updateState({ vault: MOCK_ENCRYPTION_DATA });
+
+      await keyringController.setLocked();
+
+      await keyringController.submitEncryptionKey(
+        'mockEncryptionKey',
+        'mockEncryptionSalt',
+      );
+
+      expect(keyringController.encryptor.decryptWithKey.calledOnce).toBe(true);
+      expect(keyringController.keyrings).toHaveLength(1);
+    });
+  })
+});
+```
+
+2️⃣
+
+``` typescript
+describe('KeyringController', () => {
+  describe('cacheEncryptionKey', () => {
+    it('unlocks the keyrings with valid information', async () => {
+      const keyringController = await initializeKeyringController({
+        password: 'password',
+      });
+      keyringController.cacheEncryptionKey = true;
+      const returnValue = await keyringController.encryptor.decryptWithKey();
+      const decryptWithKeyStub = sinon.stub(
+        keyringController.encryptor,
+        'decryptWithKey',
+      );
+      decryptWithKeyStub.resolves(Promise.resolve(returnValue));
+      keyringController.store.updateState({ vault: MOCK_ENCRYPTION_DATA });
+      await keyringController.setLocked();
+
+      await keyringController.submitEncryptionKey(
+        'mockEncryptionKey',
+        'mockEncryptionSalt',
+      );
+
+      expect(keyringController.encryptor.decryptWithKey.calledOnce).toBe(true);
+      expect(keyringController.keyrings).toHaveLength(1);
+    });
+  })
+});
+```
+
+### Read more
+
+- ["Arrange, act, assert"](http://xp123.com/articles/3a-arrange-act-assert/) is a simplification of the four-phase test.
+- In [behavior-driven development](https://www.agilealliance.org/glossary/bdd/), this is also called ["given, when, then"](https://martinfowler.com/bliki/GivenWhenThen.html).
+
+## Keep tests isolated
```

Comment:
This guideline is very long, and conceivably each section could be its own guideline, since there is different advice for different situations. It's also a little more philosophical than other guidelines. Let me know if anything is unclear or whether it makes sense to arrange this a different way.

Reply:
The part about determinism maybe isn't strictly needed here. We can justify test isolation without talking about that at all. I tend to think about isolation more in terms of test maintainability and readability. e.g. 

* Tests should be isolated so that they don't behave differently when `.only` is used for debugging
* Tests should be isolated so that the setup steps fully describe the test scenario to readers
* Tests should be isolated so that they don't mysteriously break when changes are made to other tests.

Plus, that statement about determinism doesn't seem right anyway. Non-deterministic tests are super useful! See fuzz testing. I prefer such tests to not be included as PR checks, instead being run on some other internval (e.g. on push to `main`, nightly, something like that), but there's still a place for non-deterministic testing.

Reply:
Hmm. I see what you're saying. I'll work on this some more.

Reply:
I tweaked the wording some here, but I like your ideas for why test isolation exists. I'll see if I can work them in tomorrow.

Reply:
With that bullet, I was thinking of exactly the same ideas you shared in this comment: https://github.com/MetaMask/contributor-docs/pull/38#discussion_r1310946590

No need to cover that here if it's covered elsewhere. Those bullets were just examples to consider